### ISSUE TITLE
torch.func.jacfwd gives incorrect Jacobian for torch.pow(torch.abs(complex), complex exponent)

### ISSUE THREAD

**Issue by shadow-orange41** [pytorch/pytorch#185639]
### 🐛 Describe the bug

For expression `torch.pow(torch.abs(z), complex_exponent)`, `torch.func.jacfwd` returns zero for a mathematically non-zero imaginary gradients, while `torch.func.jacrev` returns the correct non-zero gradients.

## Minimal producible example

```python
import torch

def complex_to_real_vec(z):
    real_part = torch.real(z)
    imag_part = torch.imag(z)
    return torch.cat([real_part, imag_part])


def make_real_wrapper(fn):
    def wrapper(coords):
        n = coords.shape[0] // 2
        real_coords = coords[:n]
        imag_coords = coords[n:]
        complex_coords = real_coords + 1j * imag_coords

        complex_res = fn(complex_coords)
        return complex_to_real_vec(complex_res)

    return wrapper


def jacobian_jacfwd(fn, x, input_size, output_size):
    x = x.detach().clone().requires_grad_(True)
    jacobian = torch.func.jacfwd(fn)(x).detach()
    return jacobian


def jacobian_jacrev(fn, x, input_size, output_size):
    x = x.detach().clone().requires_grad_(True)
    jacobian = torch.func.jacrev(fn)(x).detach()
    return jacobian


def fn(x):
    a = torch.tensor(1.0 + 1.0j, dtype=x.dtype, device=x.device)
    return torch.pow(torch.abs(x), a)


x = torch.tensor([1.0 + 1.0j], dtype=torch.complex128)

input_size = 2
output_size = 2

x_real = complex_to_real_vec(x)
real_fn = make_real_wrapper(fn)

jac_fwd = jacobian_jacfwd(real_fn, x_real, input_size, output_size)
jac_rev = jacobian_jacrev(real_fn, x_real, input_size, output_size)

print("jacfwd:")
print(jac_fwd)

print("jacrev:")
print(jac_rev)

print("difference:")
print(jac_fwd - jac_rev)
```
## Actual output

```
jacfwd:
tensor([[0.2781, 0.2781],
        [0.0000, 0.0000]], dtype=torch.float64)

jacrev:
tensor([[0.2781, 0.2781],
        [0.5471, 0.5471]], dtype=torch.float64)

difference:
tensor([[ 0.0000,  0.0000],
        [-0.5471, -0.5471]], dtype=torch.float64)
```
## Expected behavior

`torch.func.jacfwd` and `torch.func.jacrev` should agree for this real-wrapped function. The imaginary output should not have zero derivative here because: `abs(z) ** (1 + 1j) = exp((1 + 1j) * log(abs(z)))`, so the imaginary part depends on `log(abs(z))`, which depends on both `Re(z)` and `Im(z)`.

### Versions

Collecting environment information...
PyTorch version: 2.11.0+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.4 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: 16.0.6 (https://github.com/llvm/llvm-project.git 7cbf1a2591520c2491aa35339f227775f4d3adf6)
CMake version: version 4.1.0
Libc version: glibc-2.39

Python version: 3.10.20 | packaged by conda-forge | (main, Mar  5 2026, 16:42:22) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.8.0-100-generic-x86_64-with-glibc2.39
Is CUDA available: False
CUDA runtime version: 12.8.93
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No devices found.
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_adv.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_cnn.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_engines_precompiled.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_engines_runtime_compiled.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_graph.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_heuristic.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_ops.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_adv.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_cnn.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_engines_precompiled.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_engines_runtime_compiled.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_graph.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_heuristic.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_ops.so.9.8.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           52 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  384
On-line CPU(s) list:                     0-383
Vendor ID:                               AuthenticAMD
Model name:                              AMD EPYC 9654 96-Core Processor
CPU family:                              25
Model:                                   17
Thread(s) per core:                      2
Core(s) per socket:                      96
Socket(s):                               2
Stepping:                                1
Frequency boost:                         enabled
CPU(s) scaling MHz:                      48%
CPU max MHz:                             3709.3569
CPU min MHz:                             1500.0000
BogoMIPS:                                4799.86
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good amd_lbr_v2 nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba perfmon_v2 ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local user_shstk avx512_bf16 clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin cppc amd_ibpb_ret arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif x2avic v_spec_ctrl vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid overflow_recov succor smca fsrm flush_l1d debug_swap ibpb_exit_to_user
Virtualization:                          AMD-V
L1d cache:                               6 MiB (192 instances)
L1i cache:                               6 MiB (192 instances)
L2 cache:                                192 MiB (192 instances)
L3 cache:                                768 MiB (24 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-95,192-287
NUMA node1 CPU(s):                       96-191,288-383
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Mitigation; Safe RET
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; STIBP always-on; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Vulnerable: Clear CPU buffers attempted, no microcode
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Mitigation; IBPB before exit to userspace

Versions of relevant libraries:
[pip3] numpy==2.2.6
[pip3] nvidia-cublas==13.1.0.3
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cudnn-cu13==9.19.0.56
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparselt-cu13==0.8.0
[pip3] nvidia-nccl-cu13==2.28.9
[pip3] nvidia-nvjitlink==13.0.88
[pip3] nvidia-nvtx==13.0.85
[pip3] torch==2.11.0
[pip3] torchvision==0.26.0
[pip3] triton==3.6.0
[conda] numpy                     2.2.6                    pypi_0    pypi
[conda] nvidia-cublas             13.1.0.3                 pypi_0    pypi
[conda] nvidia-cuda-cupti         13.0.85                  pypi_0    pypi
[conda] nvidia-cuda-nvrtc         13.0.88                  pypi_0    pypi
[conda] nvidia-cuda-runtime       13.0.96                  pypi_0    pypi
[conda] nvidia-cudnn-cu13         9.19.0.56                pypi_0    pypi
[conda] nvidia-cufft              12.0.0.61                pypi_0    pypi
[conda] nvidia-curand             10.4.0.35                pypi_0    pypi
[conda] nvidia-cusolver           12.0.4.66                pypi_0    pypi
[conda] nvidia-cusparse           12.6.3.3                 pypi_0    pypi
[conda] nvidia-cusparselt-cu13    0.8.0                    pypi_0    pypi
[conda] nvidia-nccl-cu13          2.28.9                   pypi_0    pypi
[conda] nvidia-nvjitlink          13.0.88                  pypi_0    pypi
[conda] nvidia-nvtx               13.0.85                  pypi_0    pypi
[conda] torch                     2.11.0                   pypi_0    pypi
[conda] torchvision               0.26.0                   pypi_0    pypi
[conda] triton                    3.6.0                    pypi_0    pypi

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93 @anjali411 @dylanbespalko @mruberry @amjames @Chillee @samdow @kshitij12345

**Comment by shadow-orange41** [comment]
Similar incorrect results occur when using forward-mode autograd APIs, including `torch.func.jvp` and `torch.autograd.forward_ad`. However, `torch.autograd.functional.jvp` produces the correct results.

**Comment by praful-srinivasan-027** [comment]
## Additional debugging

I did some additional debugging and managed to reduce this further. The issue appears to be specific to `torch.pow` with a complex exponent in forward-mode AD.

I verified that the following all produce matching `jacfwd` and `jacrev` results:
- `torch.abs(x)`
- `torch.log(torch.abs(x))`
- `(1j) * torch.log(torch.abs(x))`
- `torch.exp(1j * torch.log(torch.abs(x)))`

However, `torch.pow(torch.abs(x), 1j)` still shows the discrepancy.

I also compared the JVPs of two mathematically equivalent expressions:

```python
import torch
x = torch.tensor([1.0 + 1.0j], dtype=torch.complex128)
t = torch.ones_like(x)
a = torch.tensor(1.0j, dtype=torch.complex128)

pow_primal, pow_tangent = torch.func.jvp(
    lambda z: torch.pow(torch.abs(z), a),
    (x,),
    (t,),
)
exp_primal, exp_tangent = torch.func.jvp(
    lambda z: torch.exp(a * torch.log(torch.abs(z))),
    (x,),
    (t,),
)
print(pow_primal)
print(exp_primal)
print(pow_tangent)
print(exp_tangent)
```

Output:
```
pow primal      : tensor([0.9405+0.3397j], dtype=torch.complex128)
exp-log primal  : tensor([0.9405+0.3397j], dtype=torch.complex128)
pow tangent     : tensor([-0.1698-0.0000j], dtype=torch.complex128)
exp-log tangent : tensor([-0.1698+0.4703j], dtype=torch.complex128)
```

The primal outputs match exactly, but the imaginary component of the tangent is lost in the `pow` path. I also checked that the issue disappears when using a real exponent and only appears for complex exponents.

**Comment by praful-srinivasan-027** [comment]
Further debugging suggests the issue is in the forward AD rule for `pow`.
The generated forward AD formula in `tools/autograd/derivatives.yaml` is:

```yaml
result:
  (pow_backward_self(self_t.conj(), self_p, exponent_p)
   + pow_backward_exponent(exponent_t.conj(), self_p, exponent_p, result)).conj()
```

For the reproducer above, the exponent is constant, so the JVP is coming entirely from `pow_backward_self`.
`pow_backward_self` calls:

```cpp
static Tensor handle_r_to_c(
    const Tensor& self,
    Tensor gradient_result) {

  if (!self.is_complex() && gradient_result.is_complex()) {
    return at::real(gradient_result);
  }
  return gradient_result;
}
```

I locally modified this function to temporarily return `gradient_result` instead of `at::real(gradient_result)` in the branch above and rebuilt PyTorch.
After that change:

```python
torch.func.jvp(
    lambda z: torch.pow(torch.abs(z), 1j),
    ...
)
```

produced the same tangent as:

```python
torch.func.jvp(
    lambda z: torch.exp(1j * torch.log(torch.abs(z))),
    ...
)
```

and the previously missing imaginary component reappeared.
This suggests that the imaginary component is being discarded by `handle_r_to_c` when `pow_backward_self` is reused by the generated forward-mode rule. I have not investigated what the correct fix should be, but this seems to identify the location where the discrepancy originates.

**Comment by razaaliraza** [comment]
Hi @praful-srinivasan-027 great debug work. Do you want to submit a PR for this? Thanks.

**Comment by praful-srinivasan-027** [comment]
Thanks! @razaaliraza. I'm looking into a fix now.

From the debugging so far, it seems the discrepancy originates from `handle_r_to_c` being called through `pow_backward_self`, which is reused by the generated forward AD rule. My current understanding is that `handle_r_to_c` is likely correct for reverse-mode semantics, so I'm trying to determine whether the right fix is to adjust the forward AD rule or introduce a dedicated forward-mode formula for `pow`.

I'll dig a bit further, but I wanted to check if that matches your understanding of the issue.

**Comment by praful-srinivasan-027** [comment]
I don't have a working local build yet, so I haven't been able to validate this approach.

Based on the debugging so far, I'm fairly confident that the root cause is the generated JVP rule reusing `pow_backward_self(...)`, which applies `handle_r_to_c(...)`. In the failing repro, the base is real (`torch.abs(z)`), so the projection removes the imaginary component and the forward-mode result loses the entire imaginary-output Jacobian row.

As a prototype, I introduced a `pow_backward_self_raw(...)` helper containing the existing derivative computation without the `handle_r_to_c(...)` projection. `pow_backward_self(...)` remains unchanged and simply applies the projection on top of the raw helper. I then modified the `pow.Tensor_Tensor` JVP rule to use `pow_backward_self_raw(...)` instead of `pow_backward_self(...)`.

My main question is whether this seems like the right architectural direction. The debugging suggests that the projection is the source of the discrepancy, but I'm not sure whether introducing a raw helper for the JVP path is the preferred fix, or whether `pow` should instead have a dedicated forward-mode/JVP formula.

**Comment by praful-srinivasan-027** [comment]
Hi @razaaliraza,
I've opened a draft PR with a prototype fix for this issue:
#187540 
The approach introduces `pow_backward_self_raw(...)` and uses it in the generated forward AD rule while leaving reverse-mode behavior unchanged.
I was able to reproduce the issue locally and verify that the original repro now produces matching `jacfwd` and `jacrev` results. I'd appreciate any feedback on whether this seems like the right architectural direction, or if a dedicated forward-mode formula would be preferred.
